feat: OpenTelemetry Tracing (#13899)
sveltejs/kit: documentation/docs/30-advanced/68-observability.md

---
title: Observability
---

<blockquote class="since note">

> 	<p>Available since 2.28</p>

Reviewer: 2.28 was just released https://github.com/sveltejs/kit/releases/tag/%40sveltejs%2Fkit%402.28.0
```suggestion
	<p>Available since 2.29</p>
```

</blockquote>

> [!NOTE] This feature is experimental. Expect bugs and breaking changes in minor versions (though we'll do our best to keep those to an absolute minimum). Please provide feedback!

> Sometimes, you may need to observe how your application is behaving in order to improve performance or find the root cause of a pesky bug. To help with this, SvelteKit can emit serverside [OpenTelemetry](https://opentelemetry.io) spans for the following:

Reviewer: ```suggestion
Sometimes, you may need to observe how your application is behaving to identify performance issues or the root cause of a pesky bug. To help with this, SvelteKit can emit server-side [OpenTelemetry](https://opentelemetry.io) spans for the following:
```

> - `handle`

Reviewer: ```suggestion
- `handle` hook
```

- `sequence` (`sequence`d `handle` functions will show up as children of each other and the root handle hook)
- `load` (includes univeral `load`s when they're run on the server)
- Form actions
- Remote Functions

Just telling SvelteKit to emit spans won't get you far, though -- you need to actually collect them somewhere to be able to view them. SvelteKit provides `src/tracing.server.ts` as a place to write your tracing setup and instrumentation code. It's guaranteed to be run prior to your application code being imported.

To enable both of these features, add the following to your `svelte.config.js`'s `kit` configuration:

> ```ts
> {
>   experimental: {
>     tracing: {
>       server: true,
>       serverFile: true,
>     }
>   }
> }
> ```

Reviewer: Following the example from the remote functions docs:
```suggestion
```js
/// file: svelte.config.js
export default {
	kit: {
		experimental: {
			+++tracing: {
			  server: true,
			  serverFile: true
			}+++
		}
	}
};
```
```

> > [!NOTE] Tracing -- and more significantly, tracing instrumentation -- can have a nontrivial overhead. Before you go all-in on tracing, consider whether or not you really need it, or if it might be more appropriate to turn it on in development and preview environments only.

Reviewer 2: I wonder if these should be resolve options so that you can turn them on more granularly. E.g. if a certain customer is complaining, you could turn it on for all of their employees

## Development Quickstart

> To view your first trace, you'll need to set up a local collector. We recommend [Jaeger](https://www.jaegertracing.io/docs/2.7/getting-started/), as they provide an easy-to-use quickstart command. Once your collector is running locally:

Reviewer 3: ```suggestion
To view your first trace, you'll need to set up a local collector. We recommend [Jaeger](https://www.jaegertracing.io/docs/getting-started/), as they provide an easy-to-use quickstart command. Once your collector is running locally:
```

https://www.jaegertracing.io/docs/getting-started/ redirects to latest (2.9 at time of writing this)

> - Turn on the experemental flag from above in your `svelte.config.js`

Reviewer: ```suggestion
- Turn on the experimental flag from above in your `svelte.config.js` file
```

> - Use your package manager to install `@opentelemetry/sdk-node`, `opentelemetry/auto-instrumentations-node`, `@opentelemetry/exporter-trace-otlp-proto`, and `import-in-the-middle`

Reviewer: Wonder if it's useful to have something easy to copy and paste until we have the sv add-on ready
```suggestion
- Use your package manager to install `@opentelemetry/sdk-node`, `opentelemetry/auto-instrumentations-node`, `@opentelemetry/exporter-trace-otlp-proto`, and `import-in-the-middle`
```sh
npm i @opentelemetry/sdk-node @opentelemetry/auto-instrumentations-node @opentelemetry/exporter-trace-oltp-proto import-in-the-middle
```
```

---
title: Observability
---

<blockquote class="since note">
	<p>Available since 2.29</p>
</blockquote>

> [!NOTE] This feature is experimental. Expect bugs and breaking changes in minor versions (though we'll do our best to keep those to an absolute minimum). Please provide feedback!

Sometimes, you may need to observe how your application is behaving in order to improve performance or find the root cause of a pesky bug. To help with this, SvelteKit can emit server-side [OpenTelemetry](https://opentelemetry.io) spans for the following:

> - `handle` hook (`sequence`d `handle` functions will show up as children of each other and the root handle hook) 

Reviewer 4: ```suggestion
- [`handle`](hooks#Server-hooks-handle) hook (`handle` functions running in a [`sequence`](@sveltejs-kit-hooks#sequence) will show up as children of each other and the root handle hook) 
```

- `load` functions (includes univeral `load` functions when they're run on the server)

> - Form actions
> - Remote functions

Reviewer 4: ```suggestion
- [Form actions](form-actions)
- [Remote functions](remote-functions)
```

> Just telling SvelteKit to emit spans won't get you far, though -- you need to actually collect them somewhere to be able to view them. SvelteKit provides `src/tracing.server.ts` as a place to write your tracing setup and instrumentation code. It's guaranteed to be run prior to your application code being imported.

Reviewer 4: do we need to add a note like this?

```suggestion
Just telling SvelteKit to emit spans won't get you far, though -- you need to actually collect them somewhere to be able to view them. SvelteKit provides `src/tracing.server.ts` as a place to write your tracing setup and instrumentation code. It's guaranteed to be run prior to your application code being imported, providing your deployment platform supports it and your adapter is aware of it.
```
